src/acl.erl

```diff
@@ -26,101 +26,96 @@
                       | resource | resource_regexp | resource_glob.
 
--type acl_result() :: allow | deny | term().
-
 %% Skips the domain check for the 'match => current_domain' condition
--spec match_rule(mongooseim:host_type_or_global(), rule(), jid:jid()) -> acl_result().
-match_rule(HostType, Rule, JID) ->
-    match_rule(HostType, JID#jid.lserver, Rule, JID).
+-spec match_rule(mongooseim:host_type_or_global(), rule_name(), jid:jid()) -> acl_result().
+match_rule(HostType, RuleName, JID) ->
+    match_rule(HostType, JID#jid.lserver, RuleName, JID).
 
--spec match_rule(mongooseim:host_type_or_global(), jid:lserver(), rule(), jid:jid()) ->
+-spec match_rule(mongooseim:host_type_or_global(), jid:lserver(), rule_name(), jid:jid()) ->
           acl_result().
 match_rule(_HostType, _Domain, all, _JID) ->
     allow;
 match_rule(_HostType, _Domain, none, _JID) ->
     deny;
-match_rule(HostType, Domain, Rule, JID) ->
-    match_rule(HostType, Domain, Rule, JID, deny).
+match_rule(HostType, Domain, RuleName, JID) ->
+    match_rule(HostType, Domain, RuleName, JID, deny).
 
--spec match_rule(mongooseim:host_type_or_global(), jid:lserver(), rule(), jid:jid(),
+-spec match_rule(mongooseim:host_type_or_global(), jid:lserver(), rule_name(), jid:jid(),
                  Default :: acl_result()) ->
           acl_result().
-match_rule(global, Domain, Rule, JID, Default) ->
-    case mongoose_config:lookup_opt({access, Rule, global}) of
+match_rule(HostType, Domain, RuleName, JID, Default) ->
+    case mongoose_config:lookup_opt([{access, HostType}, RuleName]) of
         {error, not_found} ->
             Default;
-        {ok, ACLs} ->
-            match_acls(ACLs, JID, global, Domain)
-    end;
-match_rule(HostType, Domain, Rule, JID, Default) ->
-    case {mongoose_config:lookup_opt({access, Rule, global}),
-          mongoose_config:lookup_opt({access, Rule, HostType})} of
-        {{error, not_found}, {error, not_found}} ->
-            Default;
-        {{error, not_found}, {ok, HostACLs}} ->
-            match_acls(HostACLs, JID, HostType, Domain);
-        {{ok, GlobalACLs}, {error, not_found}} ->
-            match_acls(GlobalACLs, JID, HostType, Domain);
-        {{ok, GlobalACLs}, {ok, HostACLs}} ->
-            match_acls(merge_acls(GlobalACLs, HostACLs), JID, HostType, Domain)
+        {ok, RuleSpec} ->
+            match_rule_clauses(RuleSpec, JID, HostType, Domain)
     end.
 
--spec merge_acls([{any(), rule()}], [{any(), rule()}]) -> [{any(), rule()}].
-merge_acls(Global, HostLocal) ->
+%% Merge host-type rules with global ones on startup
+-spec merge_access_rules([rule_clause()], [rule_clause()]) -> [rule_clause()].
+merge_access_rules(Global, HostLocal) ->
     case lists:reverse(Global) of
-        [{allow, all} | Rest] ->
-            lists:reverse(Rest) ++ HostLocal ++ [{allow, all}];
+        [#{acl := all, value := allow} = AllowAll | Rest] ->
+            lists:reverse(Rest) ++ HostLocal ++ [AllowAll];
```

Comment:
Ranting is here, **totally not related to your PR, but because of that merging logic we have today. I don't expect any rework in this PR, but in the next ones ;)**

# Part one

always find this merging logic is _a bit_ confusing :)
usually when I want to define something per host, I would just **copy** the option from the global section and modify it a bit.

And assuming that it is usually looking like 

```
max_user_sessions = [ {acl = "all", value = 10} ]
```

It is a good way to go.


But not for the access.

The current approach is a bit like defining a module, but after merging module options from global and host  sections (good, we don't do that).

# Part two

Hello, I want to configure max_user_sessions for every default host as 10, but for my specific host I want 20 (it is very good host).
Here we go:

```
[access]
  max_user_sessions = [ {acl = "all", value = 10} ]

[[host_config]]
  host = "example.com"

  [host_config.access]
    max_user_sessions = [ {acl = "all", value = 20} ]
```

Results in:

```
[#{acl => all,value => 10},#{acl => all,value => 20}] % wow, so duplicates
```

Common, it takes the value from **the global config** with this kind of merge. I would think host config rules  override the global rules (like it works for any other host_config option).
And value = 20 is just to waste computation resources, if the global rule is not matching.

The only way to configure that is to use global empty access and define different values for each host in the config.

Reply:
Changing this logic would be out of scope here. Please create a JIRA issue for it and we can groom it. It's not that simple as this logic might be already used in multiple places.